src/emulation/promises.ts

```diff
@@ -718,16 +723,43 @@ export async function readdir(
 			entries.push(entry);
 		}
 	}
-	const values: (string | Dirent)[] = [];
+
+	const values: (string | Dirent | Buffer)[] = [];
+	const hasOptions = typeof options === 'object';
+	const isRecusive = hasOptions && options?.recursive;
 	for (const entry of entries) {
-		values.push(typeof options == 'object' && options?.withFileTypes ? new Dirent(entry, await stat(join(path, entry))) : entry);
+		if (!isRecusive) {
```

Comment:
It looks like you have (simplified):
```ts
if (!isRecusive) {
	add_current_entry();
	continue;
}
add_current_entry();
```

Please de-duplicate the code by adding it before:
```ts
const fullPath = join(path, entry);
const stats = await stat(fullPath);
values.push(hasOptions && options?.withFileTypes ? new Dirent(entry, stats) : entry);

if (!isRecusive) {
	continue;
}

// ...
```

Reply:
I did this in that way because I want to avoid `stat` call before checking if it is a `withFileTypes` or not

Reply:
You can still do so, for example:
```ts
const stats = isRecusive || options?.withFileTypes ? await stat(fullPath) : null
```

Also, I recommend you normalize the options object rather than checking many times for if it is an object.